### DISCUSSION TITLE
How to: change visibility, highlight and focus of structures and shapes.

### DISCUSSION THREAD

**Discussion by hlvlad** [molstar/molstar#171] [Q&A] [answered]
Hello. Given list of `StateObjectSelector` saved from `CreateTransformer`:

1) How to change visibility of given object in list? Hide all elements except given?
2) How to highlight given object in list? Highlight two objects with different colors?
3) Focus on given object?

Thanks in advance.

**Comment by dsehnal** [comment, marked as answer]
Hi,
 
>     1. How to change visibility of given object in list? Hide all elements except given?

```ts
setSubtreeVisibility(plugin.state.data, selector.ref, true /* or false */);
```

> 
>     2. How to highlight given object in list?

```ts
PluginCommands.Interactivity.Object.Highlight(plugin, { state: plugin.state.data, ref: [sel1.ref, sel2.ref] /* or just sel1.ref if only 1 object */ });
PluginCommands.Interactivity.ClearHighlights(plugin)
```

> Highlight two objects with different colors?

The highlight color is currently a global uniform. This would have to be done using different themes for each object and calling ``plugin.build().to(selector).update({ newParams }).commit();``. Multiple ``to/update`` pairs can be added to a single state build.

>     3. Focus on given object?

One of 

```ts
plugin.managers.camera.focus*(..,)
```

functions depending on what kind of object it is. Might require getting the object data, visuals, etc depending.

**Reply by hlvlad** [reply]
Thank you.

**Reply by hlvlad** [reply]
@dsehnal, one more question. Is there is way to add custom callback to this actions? I mean, when user hovers over element in state tree or 3D Representation?

**Reply by dsehnal** [reply]
Yes, the tree uses the same command ``PluginCommands.Interactivity.Object.Highlight`` and you can just do ``PluginCommands.Interactivity.Object.Highlight.subscribe(plugin, callback)`` like say [here](https://github.com/molstar/molstar/blob/master/src/mol-plugin/behavior/static/state.ts#L106) (this also shows how to do the highlight manually).

For highlights/selections on the 3D structure you can subscribe to events in ``PluginContext.behaviors.interaction``.

**Reply by AustereTriceratops** [reply]
In `setSubtreeVisibility(plugin.state.data, selector.ref, true /* or false */);`, where is the selector coming from? How can I access it?

**Reply by dsehnal** [reply]
The selectors usually come from the state tree builder. It's something you store when you are initially creating your state.

You can also select things dynamically using `plugin.state.data.selectQ`. If you search the code base for `selectQ` you can find some examples.
